### PR TITLE
Boost: Hide generated count if the count is invalid

### LINKED ISSUE #28243: [Boost] Display issue in Critical CSS dashboard
As noted by @ajaykj, new users sometimes see `%d files generated` under the Optimize CSS Loading heading when first installed & enabled:

image

At first blush, it looks like an `undefined` or a `null` is getting in there somehow.

### PR DESCRIPTION
**Pull request by haqadn**
Fixes #28243

## Proposed changes:

* Hide generated count if the valid is invalid
* Do not show generated date if date is not valid

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
N/A

## Testing instructions:
I did not find a way to reliably reproduce this issue. So, one way to reproduce it would be to modify the generator response in https://github.com/Automattic/jetpack/blob/71b689ff874cc0e63913e1e966fc2bf24304f110/projects/plugins/boost/app/features/optimizations/critical-css/Generator.php#L73-L79 by removing the `success_count` and `updated`.

### CODE REVIEW

File: projects/plugins/boost/app/assets/src/js/pages/settings/elements/CriticalCssStatus.svelte [hunk 1]
```diff
@@ -22,16 +22,18 @@
...
-					/* translators: %d is a number of CSS Files which were successfully generated */
-					_n(
-						'%d file generated',
-						'%d files generated',
-						$criticalCssStatus.success_count,
-						'jetpack-boost'
-					),
-					$criticalCssStatus.success_count
-				)}
+				{#if typeof $criticalCssStatus.success_count === 'number' && ! isNaN( $criticalCssStatus.success_count )}
```
Comment (dilirity) [inline projects/plugins/boost/app/assets/src/js/pages/settings/elements/CriticalCssStatus.svelte]: Shouldn't

```js
! isNaN( $criticalCssStatus.success_count )
```

be enough in this case?

**Review by dilirity — COMMENTED** [review]
I couldn't reproduce this either. Having the checks there at least won't show the template string.

File: projects/plugins/boost/app/assets/src/js/pages/settings/elements/CriticalCssStatus.svelte (hunk 1 above)
Comment (thingalon) [reply projects/plugins/boost/app/assets/src/js/pages/settings/elements/CriticalCssStatus.svelte]: It depends on what value ends up in `$criticalCssStatus.success_count`.

In JavaScript, `NaN` is a specific number which is returned from invalid math operations. e.g.: `const x = 1 / 'string';` will result in `NaN` for which `isNaN` will return true. However, `isNaN( null );` returns false. It's true that `null` isn't a number, but it's not the special number called `NaN`.

**Comment by thingalon** [comment]
IMHO the problem is that we're not consistent in how we manage the Critical CSS Status object's `status` field. Let me explain it using code examples:

- In [`CloudCSSMeta`](https://github.com/Automattic/jetpack/blob/ba11d9e9d8b7530b3f35c69683b2dbfa2ed6820b/projects/plugins/boost/app/assets/src/js/pages/settings/elements/CloudCssMeta.svelte#L15) and [`CriticalCSSMeta`](https://github.com/Automattic/jetpack/blob/ba11d9e9d8b7530b3f35c69683b2dbfa2ed6820b/projects/plugins/boost/app/assets/src/js/pages/settings/elements/CriticalCssMeta.svelte#L31) we use the [`$showError` derived store](https://github.com/Automattic/jetpack/blob/ba11d9e9d8b7530b3f35c69683b2dbfa2ed6820b/projects/plugins/boost/app/assets/src/js/stores/critical-css-status.ts#L81) to check if the `status` is `error`, or if `success_count` is exactly 0.
- In [`CriticalCSSMeta`](https://github.com/Automattic/jetpack/blob/ba11d9e9d8b7530b3f35c69683b2dbfa2ed6820b/projects/plugins/boost/app/assets/src/js/pages/settings/elements/CriticalCssMeta.svelte#L9) we also branch if `status` is `'requesting'`.
- In [`CriticalCSSStatus`](https://github.com/Automattic/jetpack/blob/ba11d9e9d8b7530b3f35c69683b2dbfa2ed6820b/projects/plugins/boost/app/assets/src/js/pages/settings/elements/CriticalCssStatus.svelte#L21), we branch if `success_count` is exactly 0.

We're branching in all sorts of different places based on the value of `status` and `success_count`. It's already more complex and spread around across the code-base than I'd like.

If I read it correctly, it means that: 
- Cloud CSS will show the `%d generated` string if `success_count` is not exactly 0, and if the status is _not_ `error`.
- Critical CSS will show the `%d generated` string if the status is not `error` or `requesting`, and the status is either not `success`, or the `success_count` is not exactly 0.

I feel like these checks are expressed in weird ways, so of course we're going to fall through to showing %d under weird circumstances. I think the proposed fix just adds one more knot to an already complex net.

IMHO the correct fix is to simplify the way we check CSS status. Only expect `success_count` to be populated at all if the `status` is `success`. Make sure that the status is always `generating` when we're generating. Remove the weird branching and special cases.

**Comment by haqadn** [comment]
> Remove the weird branching and special cases.

I have done some refactoring to remove the branching. 

> Only expect `success_count` to be populated at all if the `status` is `success`. Make sure that the status is always `generating` when we're generating. 

This has resulted in a regression. We rely on `success_count` to show progress for cloud CSS, i.e. "X files generated. Jetpack Boost is generating more Critical CSS.". Thoughts? @thingalon

File: projects/plugins/boost/app/assets/src/js/elements/TimeAgo.svelte [hunk 2]
```diff
@@ -19,6 +19,8 @@
 	} );
 </script>
 
-<span title={time.toLocaleString()} class="time-ago">
-	{$label}
-</span>
+{#if ! isNaN( time.getTime() )}
```
Comment (thingalon) [inline projects/plugins/boost/app/assets/src/js/elements/TimeAgo.svelte:22]: If time is a date can getTime ever return `NaN`?

**Comment by thingalon** [comment]
> > Only expect `success_count` to be populated at all if the `status` is `success`. Make sure that the status is always `generating` when we're generating.
> 
> This has resulted in a regression. We rely on `success_count` to show progress for cloud CSS, i.e. "X files generated. Jetpack Boost is generating more Critical CSS.". Thoughts? @thingalon

Ah, I see. In that case -- I guess we need to expect `success_count` can be set if the status is regenerating or `success`, and if it's not present, treat `undefined` as 0?

File: projects/plugins/boost/app/assets/src/js/elements/TimeAgo.svelte (hunk 2 above)
Comment (haqadn) [reply projects/plugins/boost/app/assets/src/js/elements/TimeAgo.svelte:22]: Yes
```
isNaN(new Date( NaN )) && isNaN(new Date( undefined ))
```

will return true

**Comment by haqadn** [comment]
We may need to abandon this as some of the refactorings are already done on #28220. However, #28220 does not include the actual fix for the `%d` issue, So, I am not closing this yet as the fix would go on a different PR.

**Comment by thingalon** [comment]
@haqadn is this still relevant?

**Comment by haqadn** [comment]
Stale

**Pull request closed without merging**